doc/developer/design/20230209_invalid_accumulations.md

```diff
@@ -0,0 +1,78 @@
+
+Note that we would like to uniformly generate a SQL-level error in all cases above instead of outputting an erroneous negative value or crashing Materialize. The introduction of a post-processing step after aggregation that performs an appropriate [cast operation](https://github.com/MaterializeInc/materialize/blob/d28272444db09053e89eab1d568ba3a81f3da19a/src/expr/src/scalar/func.rs#L3868-L3870) would achieve this objective. Thus, implementation of the post-processing operator seems to be more natural as an MFP for category 2 errors.
+
+## Alternatives
+
+Some alternatives were discussed and ultimately discarded due to concerns regarding code maintainability or performance.
+
+### Category 1 and 3 Errors
+
+A **first alternative** is to report both the error to Sentry and also ensure that a record with a negative multiplicity is generated as part of processing. That way, upper processing layers will have a chance to observe the violation of a multiset output property and report a query-level error. However, it is not guaranteed that all query structures will result in a query-level error, since these negative multiplicities could be added up on another query path with positive multiplicities prior to output. The latter would lead to potential error reports to Sentry that are not visible to the user as SQL-level errors.
+
+A **second alternative** is to introduce a `Datum::Dummy` variant in place of the aggregated data that flags that an error has occurred. It is, however, unclear whether this would result in a panic or a more gentle query-level error. From a preliminary analysis, the result appears to be a panic, which would lead us to consider either special handling of `Datum::Dummy` or introduction of another `Datum` variant. The advantage of this approach would be higher confidence in query-level detection, as opposed to the first suggestion, but the engineering effort is less clear and the representation cleanliness is questionable.
+
+A **third alternative** is to work with `Result<_, DataflowError>` in reductions. There is, however, a concern that aggregation performance can be degraded substantially due to expansion of the size of the `enum`. This effect could be partly compensated by a better representation for `Accum` and by employing an alternative to `DataflowError` (e.g., see issue [#8086](https://github.com/MaterializeInc/materialize/issues/8066)). Still, the refactoring effort could prove to be large. Additionally, it remains unclear whether `Result` can be properly represented in arrangements.
```

Comment:
> There is, however, a concern that aggregation performance can be degraded substantially due to expansion of the size of the `enum`.

Is this still a concern now? A `Result<Row, DataflowError>` is now 8 bytes larger than a `Row`. A `Row` with an additional column would, as suggested in the design, would also increase the size in memory, although it's not clear to me by how much.

Reply:
Indeed the performance concern is much smaller after your PR that boxed `DataflowError`. But looking again at [this discussion](https://materializeinc.slack.com/archives/C02PPB50ZHS/p1674057776866419?thread_ts=1674056919.112159&cid=C02PPB50ZHS), IIUC the bigger problem is that downstream operators expect arrangements made up a `Row` key and a `Row` value. Not sure if anyone would see a simple solution to this issue? In the meantime, I am revising the text accordingly.